File: example/src/browserWindow.ts
import { browserWindow } from "vokex";

import { log, clear } from './utils'

log("页面已加载");

let mainWindow: any | null = null;

document.getElementById("btn-win-create")?.addEventListener("click", async () => {
  clear();
  log("=== browserWindow.create() ===");
  try {
    const newWin = await browserWindow.create({
      title: `新窗口 - ${new Date().toLocaleTimeString()}`,
      width: 600,
      height: 400,
    });
    log(`✅ 新窗口已创建，ID: ${newWin.getId()}`);
    const allWindows = await browserWindow.getAll();
    log(`当前共有 ${allWindows.length} 个窗口`);
  } catch (error: any) {
    log(`❌ 错误: ${error.message}`);
  }
});

document.getElementById("btn-win-set-min-size")?.addEventListener("click", async () => {
  clear();
  log("=== win.setMinimumSize() ===");
  try {
    if (!mainWindow) {
      mainWindow = await browserWindow.getCurrentWindow();
    }
    if (mainWindow) {
      await mainWindow.setMinimumSize(400, 300);
      log("✅ 最小窗口尺寸已设置为: 400 x 300");
      log("提示: 尝试拖动窗口边缘缩小，无法小于该尺寸");
      log("提示: 10秒后取消限制");
      setTimeout(async () => {
        await mainWindow?.setMinimumSize(0, 0);
        log("✅ 最小尺寸限制已取消");
      }, 10000);
    } else {
      log("⚠️ 未找到主窗口");
    }
  } catch (error: any) {
    log(`❌ 错误: ${error.message}`);
  }
});

document.getElementById("btn-win-set-bg-color")?.addEventListener("click", async () => {
  clear();
  log("=== win.setBackgroundColor() ===");
  try {
    if (!mainWindow) {
      mainWindow = await browserWindow.getCurrentWindow();
    }
    if (mainWindow) {
      const color = "#ff6600";
      await mainWindow.setBackgroundColor(color);
      log(`✅ 背景色已设置为: ${color}`);
      log("提示: 3秒后恢复白色");
      setTimeout(async () => {
        await mainWindow?.setBackgroundColor("#ffffff");
        log("✅ 背景色已恢复");
      }, 3000);
    } else {
      log("⚠️ 未找到主窗口");
    }
  } catch (error: any) {
    log(`❌ 错误: ${error.message}`);
  }
});

document.getElementById("btn-win-skip-taskbar")?.addEventListener("click", async () => {
  clear();
  log("=== win.setSkipTaskbar() [Windows] ===");
  try {
    if (!mainWindow) {
      mainWindow = await browserWindow.getCurrentWindow();
    }
    if (mainWindow) {
      await mainWindow.setSkipTaskbar(true);
      log("✅ 窗口已从任务栏隐藏");
      log("提示: 3秒后恢复显示");
      setTimeout(async () => {
        await mainWindow?.setSkipTaskbar(false);
        log("✅ 窗口已恢复到任务栏");
      }, 3000);
    } else {
      log("⚠️ 未找到主窗口");
    }
  } catch (error: any) {
    log(`❌ 错误: ${error.message}`);
  }
});

document.getElementById("btn-win-blur")?.addEventListener("click", async () => {
  clear();
  log("=== win.blur() ===");
  try {
    if (!mainWindow) {
      mainWindow = await browserWindow.getCurrentWindow();
    }
    if (mainWindow) {
      await mainWindow.blur();
      log("✅ 窗口已失焦（最小化后恢复）");
    } else {
      log("⚠️ 未找到主窗口");
    }
  } catch (error: any) {
    log(`❌ 错误: ${error.message}`);
  }
});
document.getElementById("btn-win-get-main")?.addEventListener("click", async () => {
  clear();
  log("=== browserWindow.getCurrentWindow() ===");
  try {
    const mainWindow = await browserWindow.getWindow(1);
    if (mainWindow) {
      log(`✅ 获取到主窗口，ID: ${mainWindow.getId()}`);
      const title = await mainWindow.getTitle();
      log(`窗口标题: ${title}`);
    } else {
      log("⚠️ 未找到主窗口");
    }
  } catch (error: any) {
    log(`❌ 错误: ${error.message}`);
  }
});

document.getElementById("btn-win-get-all")?.addEventListener("click", async () => {
  clear();
  log("=== browserWindow.getAll() ===");
  try {
    const windows = await browserWindow.getAll();
    log(`共有 ${windows.length} 个窗口:`);
    windows.forEach((win, i) => {
      log(`  [${i}] ID: ${win.id}, 标题: ${win.title}`);
      log(`       大小: ${win.width}x${win.height}, 位置: (${win.x}, ${win.y})`);
      log(`       最大化: ${win.is_maximized}, 最小化: ${win.is_minimized}, 全屏: ${win.is_fullscreen}`);
      log(`       聚焦: ${win.is_focused}, 可见: ${win.is_visible}`);
    });
  } catch (error: any) {
    log(`❌ 错误: ${error.message}`);
  }
});

document.getElementById("btn-win-get-focused")?.addEventListener("click", async () => {
  clear();
  log("=== browserWindow.getFocused() ===");
  try {
    const focused = await browserWindow.getFocused();
    if (focused) {
      log(`✅ 当前聚焦窗口 ID: ${focused.id}`);
      log(`标题: ${focused.title}`);
    } else {
      log("⚠️ 没有窗口处于聚焦状态");
    }
  } catch (error: any) {
    log(`❌ 错误: ${error.message}`);
  }
});

document.getElementById("btn-win-set-title")?.addEventListener("click", async () => {
  clear();
  log("=== win.setTitle() ===");
  try {
    if (!mainWindow) {
      mainWindow = await browserWindow.getCurrentWindow();
    }
    if (mainWindow) {
      const newTitle = `Vokex - ${new Date().toLocaleTimeString()}`;
      await mainWindow.setTitle(newTitle);
      log(`✅ 窗口标题已设置为: ${newTitle}`);
      const title = await mainWindow.getTitle();
      log(`当前标题: ${title}`);
    } else {
      log("⚠️ 未找到主窗口");
    }
  } catch (error: any) {
    log(`❌ 错误: ${error.message}`);
  }
});

document.getElementById("btn-win-get-size")?.addEventListener("click", async () => {
  clear();
  log("=== win.getSize() ===");
  try {
    if (!mainWindow) {
      mainWindow = await browserWindow.getCurrentWindow();
    }
    if (mainWindow) {
      const size = await mainWindow.getSize();
      log(`窗口大小: ${size[0]} x ${size[1]}`);
      const pos = await mainWindow.getPosition();
      log(`窗口位置: (${pos[0]}, ${pos[1]})`);
    } else {
      log("⚠️ 未找到主窗口");
    }
  } catch (error: any) {
    log(`❌ 错误: ${error.message}`);
  }
});

document.getElementById("btn-win-set-size")?.addEventListener("click", async () => {
  clear();
  log("=== win.setSize() ===");
  try {
    if (!mainWindow) {
      mainWindow = await browserWindow.getCurrentWindow();
    }
    if (mainWindow) {
      const width = 800 + Math.floor(Math.random() * 200);
      const height = 600 + Math.floor(Math.random() * 200);
      await mainWindow.setSize(width, height);
      log(`✅ 窗口大小已设置为: ${width} x ${height}`);
      const size = await mainWindow.getSize();
      log(`实际大小: ${size[0]} x ${size[1]}`);
    } else {
      log("⚠️ 未找到主窗口");
    }
  } catch (error: any) {
    log(`❌ 错误: ${error.message}`);
  }
});

document.getElementById("btn-win-center")?.addEventListener("click", async () => {
  clear();
  log("=== win.center() ===");
  try {
    if (!mainWindow) {
      mainWindow = await browserWindow.getCurrentWindow();
    }
    if (mainWindow) {
      await mainWindow.center();
      log("✅ 窗口已居中");
      const pos = await mainWindow.getPosition();
      log(`新位置: (${pos[0]}, ${pos[1]})`);
    } else {
      log("⚠️ 未找到主窗口");
    }
  } catch (error: any) {
    log(`❌ 错误: ${error.message}`);
  }
});

document.getElementById("btn-win-minimize")?.addEventListener("click", async () => {
  clear();
  log("=== win.minimize() ===");
  try {
    if (!mainWindow) {
      mainWindow = await browserWindow.getCurrentWindow();
    }
    if (mainWindow) {
      await mainWindow.minimize();
      log("✅ 窗口已最小化");
      const isMinimized = await mainWindow.isMinimized();
      log(`最小化状态: ${isMinimized}`);
    } else {
      log("⚠️ 未找到主窗口");
    }
  } catch (error: any) {
    log(`❌ 错误: ${error.message}`);
  }
});

document.getElementById("btn-win-maximize")?.addEventListener("click", async () => {
  clear();
  log("=== win.maximize() ===");
  try {
    if (!mainWindow) {
      mainWindow = await browserWindow.getCurrentWindow();
    }
    if (mainWindow) {
      await mainWindow.maximize();
      log("✅ 窗口已最大化");
      const isMaximized = await mainWindow.isMaximized();
      log(`最大化状态: ${isMaximized}`);
    } else {
      log("⚠️ 未找到主窗口");
    }
  } catch (error: any) {
    log(`❌ 错误: ${error.message}`);
  }
});

document.getElementById("btn-win-restore")?.addEventListener("click", async () => {
  clear();
  log("=== win.restore() ===");
  try {
    if (!mainWindow) {
      mainWindow = await browserWindow.getCurrentWindow();
    }
    if (mainWindow) {
      await mainWindow.restore();
      log("✅ 窗口已恢复");
      const isMinimized = await mainWindow.isMinimized();
      const isMaximized = await mainWindow.isMaximized();
      log(`最小化: ${isMinimized}, 最大化: ${isMaximized}`);
    } else {
      log("⚠️ 未找到主窗口");
    }
  } catch (error: any) {
    log(`❌ 错误: ${error.message}`);
  }
});

document.getElementById("btn-win-fullscreen")?.addEventListener("click", async () => {
  clear();
  log("=== win.setFullScreen() ===");
  try {
    if (!mainWindow) {
      mainWindow = await browserWindow.getCurrentWindow();
    }
    if (mainWindow) {
      const isFullScreen = await mainWindow.isFullScreen();
      log(`当前全屏状态: ${isFullScreen}`);
      await mainWindow.setFullScreen(!isFullScreen);
      log(`✅ 全屏状态已切换为: ${!isFullScreen}`);
    } else {
      log("⚠️ 未找到主窗口");
    }
  } catch (error: any) {
    log(`❌ 错误: ${error.message}`);
  }
});

document.getElementById("btn-win-always-on-top")?.addEventListener("click", async () => {
  clear();
  log("=== win.setAlwaysOnTop() ===");
  try {
    if (!mainWindow) {
      mainWindow = await browserWindow.getCurrentWindow();
    }
    if (mainWindow) {
      await mainWindow.setAlwaysOnTop(true);
      log("✅ 窗口已设置为置顶");
      log("提示: 3秒后取消置顶");
      setTimeout(async () => {
        await mainWindow?.setAlwaysOnTop(false);
        log("✅ 置顶已取消");
      }, 3000);
    } else {
      log("⚠️ 未找到主窗口");
    }
  } catch (error: any) {
    log(`❌ 错误: ${error.message}`);
  }
});

document.getElementById("btn-win-set-position")?.addEventListener("click", async () => {
  clear();
  log("=== win.setPosition() ===");
  try {
    if (!mainWindow) {
      mainWindow = await browserWindow.getCurrentWindow();
    }
    if (mainWindow) {
      const x = 100 + Math.floor(Math.random() * 200);
      const y = 100 + Math.floor(Math.random() * 200);
      await mainWindow.setPosition(x, y);
      log(`✅ 窗口位置已设置为: (${x}, ${y})`);
      const pos = await mainWindow.getPosition();
      log(`实际位置: (${pos[0]}, ${pos[1]})`);
    } else {
      log("⚠️ 未找到主窗口");
    }
  } catch (error: any) {
    log(`❌ 错误: ${error.message}`);
  }
});

document.getElementById("btn-win-set-resizable")?.addEventListener("click", async () => {
  clear();
  log("=== win.setResizable() ===");
  try {
    if (!mainWindow) {
      mainWindow = await browserWindow.getCurrentWindow();
    }
    if (mainWindow) {
      await mainWindow.setResizable(false);
      log("✅ 窗口已设置为不可调整大小");
      log("提示: 10秒后恢复可调整大小");
      setTimeout(async () => {
        await mainWindow?.setResizable(true);
        log("✅ 窗口已恢复可调整大小");
      }, 10000);
    } else {
      log("⚠️ 未找到主窗口");
    }
  } catch (error: any) {
    log(`❌ 错误: ${error.message}`);
  }
});

document.getElementById("btn-win-set-opacity")?.addEventListener("click", async () => {
  clear();
  log("=== win.setOpacity() [Windows] ===");
  try {
    if (!mainWindow) {
      mainWindow = await browserWindow.getCurrentWindow();
    }
    if (mainWindow) {
      const opacity = 0.5;
      await mainWindow.setOpacity(opacity);
      log(`✅ 窗口透明度已设置为: ${opacity}`);
      log("提示: 3秒后恢复不透明");
      setTimeout(async () => {
        await mainWindow?.setOpacity(1.0);
        log("✅ 窗口已恢复不透明");
      }, 3000);
    } else {
      log("⚠️ 未找到主窗口");
    }
  } catch (error: any) {
    log(`❌ 错误: ${error.message}`);
    log("提示: setOpacity 仅在 Windows 和 macOS 上可用");
  }
});

document.getElementById("btn-win-set-progress")?.addEventListener("click", async () => {
  clear();
  log("=== win.setProgressBar() [Windows] ===");
  try {
    if (!mainWindow) {
      mainWindow = await browserWindow.getCurrentWindow();
    }
    if (mainWindow) {
      const progress = 0.5;
      await mainWindow.setProgressBar(progress);
      log(`✅ 任务栏进度条已设置为: ${progress * 100}%`);
      log("提示: 3秒后隐藏进度条");
      setTimeout(async () => {
        await mainWindow?.setProgressBar(-1);
        log("✅ 进度条已隐藏");
      }, 3000);
    } else {
      log("⚠️ 未找到主窗口");
    }
  } catch (error: any) {
    log(`❌ 错误: ${error.message}`);
    log("提示: setProgressBar 仅在 Windows 和 macOS 上可用");
  }
});

document.getElementById("btn-win-reload")?.addEventListener("click", async () => {
  clear();
  log("=== win.reload() ===");
  try {
    if (!mainWindow) {
      mainWindow = await browserWindow.getCurrentWindow();
    }
    if (mainWindow) {
      await mainWindow.reload();
      log("✅ 窗口已重新加载");
    } else {
      log("⚠️ 未找到主窗口");
    }
  } catch (error: any) {
    log(`❌ 错误: ${error.message}`);
  }
});

document.getElementById("btn-win-hide-show")?.addEventListener("click", async () => {
  clear();
  log("=== win.hide() / win.show() ===");
  try {
    if (!mainWindow) {
      mainWindow = await browserWindow.getCurrentWindow();
    }
    if (mainWindow) {
      await mainWindow.hide();
      log("✅ 窗口已隐藏");
      log("提示: 2秒后显示窗口");
      setTimeout(async () => {
        await mainWindow?.show();
        log("✅ 窗口已显示");
      }, 2000);
    } else {
      log("⚠️ 未找到主窗口");
    }
  } catch (error: any) {
    log(`❌ 错误: ${error.message}`);
  }
});

document.getElementById("btn-win-flash-taskbar")?.addEventListener("click", async () => {
  clear();
  log("=== win.flashTaskbar() [Windows] ===");
  try {
    if (!mainWindow) {
      mainWindow = await browserWindow.getCurrentWindow();
    }
    if (mainWindow) {
      await mainWindow.flashTaskbar(true);
      log("✅ 任务栏图标正在闪烁（5次）");
      log("提示: 窗口未激活时任务栏会橙色闪烁");
    } else {
      log("⚠️ 未找到主窗口");
    }
  } catch (error: any) {
    log(`❌ 错误: ${error.message}`);
  }
});

File: shell/src/apis/browser_window.rs
use serde_json::{json, Value};
use tao::dpi::{LogicalPosition, LogicalSize};

#[cfg(target_os = "windows")]
#[repr(C)]
struct FLASHWINFO {
    cb_size: u32,
    hwnd: isize,
    dw_flags: u32,
    u_count: u32,
    dw_timeout: u32,
}

#[cfg(target_os = "windows")]
extern "system" {
    fn GetSystemMetrics(index: i32) -> i32;
    fn GetWindowLongW(hwnd: isize, index: i32) -> i32;
    fn SetWindowLongW(hwnd: isize, index: i32, value: i32) -> i32;
    fn SetLayeredWindowAttributes(hwnd: isize, crkey: u32, balpha: u8, flags: u32) -> i32;
    fn FlashWindowEx(info: *const FLASHWINFO) -> i32;
}


/// browserWindow 模块 API 处理
/// 注意：browserWindow.create 在 ipc.rs 中被拦截，直接走 CreateWindow 事件
pub fn handle(method: &str, params: &Value) -> Result<Value, String> {
    match method {
        "browserWindow.getAll" => get_all(params),
        "browserWindow.getFocused" => get_focused(params),
        "browserWindow.getById" => get_by_id(params),
        "browserWindow.close" => with_window(params, |w| { w.window.set_visible(false); Ok(json!(true)) }),
        "browserWindow.show" => with_window(params, |w| { w.window.set_visible(true); Ok(json!(true)) }),
        "browserWindow.hide" => with_window(params, |w| { w.window.set_visible(false); Ok(json!(true)) }),
        "browserWindow.minimize" => with_window(params, |w| { w.window.set_minimized(true); Ok(json!(true)) }),
        "browserWindow.maximize" => with_window(params, |w| { w.window.set_maximized(true); Ok(json!(true)) }),
        "browserWindow.unmaximize" => with_window(params, |w| { w.window.set_maximized(false); Ok(json!(true)) }),
        "browserWindow.restore" => with_window(params, |w| { w.window.set_minimized(false); w.window.set_maximized(false); Ok(json!(true)) }),
        "browserWindow.focus" => with_window(params, |w| { w.window.set_focus(); Ok(json!(true)) }),
        "browserWindow.isMaximized" => with_window(params, |w| Ok(json!(w.window.is_maximized()))),
        "browserWindow.isMinimized" => with_window(params, |w| Ok(json!(w.window.is_minimized()))),
        "browserWindow.isFullScreen" => with_window(params, |w| Ok(json!(w.window.fullscreen().is_some()))),
        "browserWindow.isFocused" => with_window(params, |w| Ok(json!(w.window.is_focused()))),
        "browserWindow.isVisible" => with_window(params, |w| Ok(json!(w.window.is_visible()))),
        "browserWindow.isResizable" => with_window(params, |w| Ok(json!(w.window.is_resizable()))),
        "browserWindow.setFullScreen" => with_window(params, |w| {
            let flag = params.get("flag").and_then(|v| v.as_bool()).unwrap_or(true);
            if flag {
                w.window.set_fullscreen(Some(tao::window::Fullscreen::Borderless(None)));
            } else {
                w.window.set_fullscreen(None);
            }
            Ok(json!(true))
        }),
        "browserWindow.setTitle" => with_window(params, |w| {
            let title = params.get("title").and_then(|v| v.as_str()).unwrap_or("");
            w.window.set_title(title);
            Ok(json!(true))
        }),
        "browserWindow.getTitle" => with_window(params, |w| Ok(json!(w.window.title()))),
        "browserWindow.setSize" => with_window(params, |w| {
            let width = params.get("width").and_then(|v| v.as_f64()).unwrap_or(800.0);
            let height = params.get("height").and_then(|v| v.as_f64()).unwrap_or(600.0);
            w.window.set_inner_size(LogicalSize::new(width, height));
            Ok(json!(true))
        }),
        "browserWindow.getSize" => with_window(params, |w| {
            let size = w.window.inner_size();
            Ok(json!([size.width, size.height]))
        }),
        "browserWindow.setMinimumSize" => with_window(params, |w| {
            let width = params.get("width").and_then(|v| v.as_f64()).unwrap_or(0.0);
            let height = params.get("height").and_then(|v| v.as_f64()).unwrap_or(0.0);
            w.window.set_min_inner_size(Some(LogicalSize::new(width, height)));
            Ok(json!(true))
        }),
        "browserWindow.setMaximumSize" => with_window(params, |w| {
            let width = params.get("width").and_then(|v| v.as_f64()).unwrap_or(0.0);
            let height = params.get("height").and_then(|v| v.as_f64()).unwrap_or(0.0);
            w.window.set_max_inner_size(Some(LogicalSize::new(width, height)));
            Ok(json!(true))
        }),
        "browserWindow.setResizable" => with_window(params, |w| {
            let flag = params.get("flag").and_then(|v| v.as_bool()).unwrap_or(true);
            w.window.set_resizable(flag);
            Ok(json!(true))
        }),
        "browserWindow.setAlwaysOnTop" => with_window(params, |w| {
            let flag = params.get("flag").and_then(|v| v.as_bool()).unwrap_or(false);
            w.window.set_always_on_top(flag);
            Ok(json!(true))
        }),
        "browserWindow.setPosition" => with_window(params, |w| {
            let x = params.get("x").and_then(|v| v.as_f64()).unwrap_or(0.0);
            let y = params.get("y").and_then(|v| v.as_f64()).unwrap_or(0.0);
            w.window.set_outer_position(LogicalPosition::new(x, y));
            Ok(json!(true))
        }),
        "browserWindow.getPosition" => with_window(params, |w| {
            let pos = w.window.outer_position().map_err(|e| format!("{}", e))?;
            Ok(json!([pos.x, pos.y]))
        }),
        "browserWindow.center" => with_window(params, |w| {
            if let Some(monitor) = w.window.current_monitor() {
                let screen_size = monitor.size();
                let screen_pos = monitor.position();
                let win_size = w.window.outer_size();
                let x = screen_pos.x + ((screen_size.width as i32 - win_size.width as i32) / 2);
                let y = screen_pos.y + ((screen_size.height as i32 - win_size.height as i32) / 2);
                w.window.set_outer_position(tao::dpi::PhysicalPosition::new(x, y));
            }
            Ok(json!(true))
        }),
        "browserWindow.setOpacity" => with_window(params, |w| {
            let opacity = params.get("opacity").and_then(|v| v.as_f64()).unwrap_or(1.0);
            #[cfg(target_os = "windows")]
            {
                use raw_window_handle::HasWindowHandle;
                if let Ok(handle) = w.window.window_handle() {
                    let raw = handle.as_raw();
                    let hwnd = match raw {
                        raw_window_handle::RawWindowHandle::Win32(h) => h.hwnd.get() as isize,
                        _ => return Err("Unsupported platform".to_string()),
                    };
                    const GWL_EXSTYLE: i32 = -20;
                    const WS_EX_LAYERED: i32 = 0x00080000;
                    const LWA_ALPHA: u32 = 0x02;
                    unsafe {
                        let ex_style = GetWindowLongW(hwnd, GWL_EXSTYLE);
                        SetWindowLongW(hwnd, GWL_EXSTYLE, ex_style | WS_EX_LAYERED);
                        let alpha = if opacity <= 0.0 { 0u8 } else if opacity >= 1.0 { 255u8 } else { (opacity * 255.0) as u8 };
                        SetLayeredWindowAttributes(hwnd, 0, alpha, LWA_ALPHA);
                    }
                }
            }
            Ok(json!(true))
        }),
        "browserWindow.setBackgroundColor" => with_window(params, |w| {
            let color_str = params.get("color").and_then(|v| v.as_str()).unwrap_or("#ffffff");
            let (r, g, b, a) = parse_color(color_str);
            w.window.set_background_color(Some((r, g, b, a)));
            Ok(json!(true))
        }),
        "browserWindow.reload" => with_window(params, |w| {
            let _ = w.webview.reload();
            Ok(json!(true))
        }),
        "browserWindow.loadURL" => with_window(params, |w| {
            let url = params.get("url").and_then(|v| v.as_str()).unwrap_or("");
            let _ = w.webview.load_url(url);
            Ok(json!(true))
        }),
        "browserWindow.loadFile" => with_window(params, |w| {
            let path = params.get("path").and_then(|v| v.as_str()).unwrap_or("");
            let url = format!("file://{}", path.replace('\\', "/"));
            let _ = w.webview.load_url(&url);
            Ok(json!(true))
        }),
        "browserWindow.setProgressBar" => with_window(params, |w| {
            let progress = params.get("progress").and_then(|v| v.as_f64()).unwrap_or(-1.0);
            let state = if progress < 0.0 {
                tao::window::ProgressBarState {
                    state: Some(tao::window::ProgressState::None),
                    progress: None,
                    desktop_filename: None,
                }
            } else if progress >= 1.0 {
                tao::window::ProgressBarState {
                    state: Some(tao::window::ProgressState::Paused),
                    progress: Some(100),
                    desktop_filename: None,
                }
            } else {
                tao::window::ProgressBarState {
                    state: Some(tao::window::ProgressState::Normal),
                    progress: Some((progress * 100.0) as u64),
                    desktop_filename: None,
                }
            };
            w.window.set_progress_bar(state);
            Ok(json!(true))
        }),
        "browserWindow.setSkipTaskbar" => with_window(params, |w| {
            let flag = params.get("flag").and_then(|v| v.as_bool()).unwrap_or(true);
            #[cfg(target_os = "windows")]
            {
                use raw_window_handle::HasWindowHandle;
                if let Ok(handle) = w.window.window_handle() {
                    let raw = handle.as_raw();
                    let hwnd = match raw {
                        raw_window_handle::RawWindowHandle::Win32(h) => h.hwnd.get() as isize,
                        _ => return Err("Unsupported platform".to_string()),
                    };
                    const GWL_EXSTYLE: i32 = -20;
                    const WS_EX_APPWINDOW: i32 = 0x00040000;
                    const WS_EX_TOOLWINDOW: i32 = 0x00000080;
                    unsafe {
                        let ex_style = GetWindowLongW(hwnd, GWL_EXSTYLE);
                        if flag {
                            // 隐藏：去掉 APPWINDOW，加上 TOOLWINDOW
                            SetWindowLongW(hwnd, GWL_EXSTYLE, (ex_style & !WS_EX_APPWINDOW) | WS_EX_TOOLWINDOW);
                        } else {
                            // 显示：加上 APPWINDOW，去掉 TOOLWINDOW
                            SetWindowLongW(hwnd, GWL_EXSTYLE, (ex_style | WS_EX_APPWINDOW) & !WS_EX_TOOLWINDOW);
                        }
                    }
                }
            }
            Ok(json!(true))
        }),
        "browserWindow.capturePage" => with_window(params, |_| {
            Err("capturePage not yet implemented".to_string())
        }),
        "browserWindow.setIcon" => with_window(params, |w| {
            let icon_path = params.get("icon").and_then(|v| v.as_str()).unwrap_or("");
            if icon_path.is_empty() {
                w.window.set_window_icon(None);
            } else {
                let image = crate::utils::load_image(icon_path);
                w.window.set_window_icon(image);
            }
            Ok(json!(true))
        }),
        "browserWindow.blur" => with_window(params, |w| {
            w.window.set_minimized(true);
            w.window.set_minimized(false);
            Ok(json!(true))
        }),
        "browserWindow.flashTaskbar" => with_window(params, |w| {
            #[cfg(target_os = "windows")]
            {
                use raw_window_handle::HasWindowHandle;
                if let Ok(handle) = w.window.window_handle() {
                    let raw = handle.as_raw();
                    let hwnd = match raw {
                        raw_window_handle::RawWindowHandle::Win32(h) => h.hwnd.get() as isize,
                        _ => return Err("Unsupported platform".to_string()),
                    };
                    let flag = params.get("flag").and_then(|v| v.as_bool()).unwrap_or(true);
                    // FLASHW_ALL = 3 (任务栏 + 标题栏), FLASHW_TIMERNOFG = 12 (未激活时闪烁)
                    let dw_flags = if flag { 3u32 } else { 0u32 };
                    unsafe {
                        let info = FLASHWINFO {
                            cb_size: std::mem::size_of::<FLASHWINFO>() as u32,
                            hwnd,
                            dw_flags,
                            u_count: if flag { 5 } else { 0 },
                            dw_timeout: 0,
                        };
                        FlashWindowEx(&info);
                    }
                }
            }
            Ok(json!(true))
        }),
        _ => Err(format!("Unknown method: {}", method)),
    }
}

/// 从 params 中提取 window_id 并执行操作
fn with_window<F>(params: &Value, f: F) -> Result<Value, String>
where
    F: FnOnce(&crate::window_manager::WindowEntry) -> Result<Value, String>,
{
    let id = params.get("id").and_then(|v| v.as_u64()).unwrap_or(0) as u32;
    crate::window_manager::MANAGER.with(|m| {
        let manager = m.borrow();
        let entry = manager.get(id).ok_or_else(|| format!("Window {} not found", id))?;
        f(entry)
    })
}

fn get_all(_params: &Value) -> Result<Value, String> {
    let ids = crate::window_manager::get_all_ids();
    let mut windows = Vec::new();
    for id in ids {
        let info = crate::window_manager::MANAGER.with(|m| {
            let manager = m.borrow();
            manager.get(id).map(|entry| entry.get_info(id))
        });
        if let Some(info) = info {
            windows.push(serde_json::to_value(info).unwrap_or_default());
        }
    }
    Ok(json!(windows))
}

fn get_focused(_params: &Value) -> Result<Value, String> {
    let ids = crate::window_manager::get_all_ids();
    for id in ids {
        let found = crate::window_manager::MANAGER.with(|m| {
            let manager = m.borrow();
            if let Some(entry) = manager.get(id) {
                if entry.window.is_focused() {
                    return Some(entry.get_info(id));
                }
            }
            None
        });
        if let Some(info) = found {
            return Ok(serde_json::to_value(info).unwrap_or_default());
        }
    }
    Ok(json!(null))
}

fn get_by_id(params: &Value) -> Result<Value, String> {
    let id = params.get("id").and_then(|v| v.as_u64()).unwrap_or(0) as u32;
    crate::window_manager::MANAGER.with(|m| {
        let manager = m.borrow();
        match manager.get(id) {
            Some(entry) => Ok(serde_json::to_value(entry.get_info(id)).unwrap_or_default()),
            None => Ok(json!(null)),
        }
    })
}

/// 解析颜色字符串为 RGBA 元组（支持 #RRGGBB 和 #RRGGBBAA 格式）
fn parse_color(color: &str) -> (u8, u8, u8, u8) {
    let hex = color.trim_start_matches('#');
    let r = u8::from_str_radix(&hex[0..2.min(hex.len())], 16).unwrap_or(255);
    let g = u8::from_str_radix(&hex[2..4.min(hex.len())], 16).unwrap_or(255);
    let b = u8::from_str_radix(&hex[4..6.min(hex.len())], 16).unwrap_or(255);
    let a = if hex.len() >= 8 {
        u8::from_str_radix(&hex[6..8], 16).unwrap_or(255)
    } else {
        255
    };
    (r, g, b, a)
}
